jgrapht-core/src/main/java/org/jgrapht/alg/spanning/AhujaOrlinSharmaCapacitatedMinimumSpanningTree.java

```diff
@@ -0,0 +1,1336 @@
+                improvementGraph.removeEdge(v1, v2);
+            }
+        }
+
+        /**
+         * Calculates the maximum demand over all new subtrees induced by the minimum spanning tree <code>spanningTree</code>.
+         * A spanning tree induces more than one subset in the partition if the root vertex of the base graph connects more than one subtree of the spanning tree.
+         *
+         * @param vertexSubset the vertex subset <code>spanning Tree is defined on</code>
+         * @param spanningTree the spanning tree
+         * @param totalDemand  the total demand of the whole spanning tree
+         * @return the maximum demand over all new subtrees induced by the minimum spanning tree <code>spanningTree</code>
+         */
+        public double calculateMaximumDemandOfSubtrees(Set<V> vertexSubset, SpanningTreeAlgorithm.SpanningTree<E> spanningTree, double totalDemand) {
+
+            Graph<V, E> spanningTreeGraph = new AsSubgraph<>(graph, vertexSubset, spanningTree.getEdges());
+
+            /*
+             * The subtree does not evolve to more than 1 partition subsets, thus, we can return the total demand.
+             */
+            int degreeOfRoot = spanningTreeGraph.degreeOf(root);
+            if (degreeOfRoot == 1) {
+                return totalDemand;
+            }
+
+            double maximumDemand = 0;
+
+            DepthFirstIterator<V, E> depthFirstIterator = new DepthFirstIterator<>(spanningTreeGraph, root);
+            if (depthFirstIterator.hasNext()) {
+                depthFirstIterator.next();
+            }
+
+            int numberOfRootEdgesExplored = 0;
+
+            double exploredVerticesDemand = 0;
+            double currentDemand = 0;
+
+            while (depthFirstIterator.hasNext()) {
+                V next = depthFirstIterator.next();
+
+                // exploring new subtree
+                if (spanningTreeGraph.containsEdge(root, next)) {
+
+                    exploredVerticesDemand += currentDemand;
+
+                    if (maximumDemand < currentDemand) {
+                        maximumDemand = currentDemand;
+                    }
+
+                    // we can stop the exploration
+                    if (maximumDemand >= 0.5 * totalDemand || exploredVerticesDemand + maximumDemand >= totalDemand) {
+                        return maximumDemand;
+                    }
+
+                    // we can stop the exploration, all subtrees but one are explored
+                    if (numberOfRootEdgesExplored + 1 == degreeOfRoot) {
+                        return Math.max(maximumDemand, totalDemand - exploredVerticesDemand);
+                    }
+
+                    numberOfRootEdgesExplored++;
+
+                    currentDemand = 0;
+                }
+
+                currentDemand += demands.get(next);
+            }
+
+            return maximumDemand;
+        }
+
+        /**
+         * Returns the mapping that is used in the valid cycle detecti
```

Comment:
Usually custom implementations of Map extend the skeleton implementation `AbstractMap`.